File: engine/src/loss.rs
//! Loss functions, both as eager scalars (for inference / metrics) and as
//! autograd nodes (for training).

use crate::activations::softmax_rows;
use crate::autograd::{Tape, Value};
use crate::tensor::Tensor;
use serde::{Deserialize, Serialize};

#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum Loss {
    /// Mean-squared error. Pair with `Activation::Identity`, `Sigmoid`, or `Tanh`.
    MeanSquaredError,
    /// Softmax + categorical cross-entropy. The model's last layer should be
    /// `Activation::Identity` — softmax is applied internally for numerical
    /// stability.
    CrossEntropy,
}

impl Loss {
    pub fn name(&self) -> &'static str {
        match self {
            Loss::MeanSquaredError => "MeanSquaredError",
            Loss::CrossEntropy => "CrossEntropy",
        }
    }

    pub fn all() -> &'static [Loss] {
        &[Loss::MeanSquaredError, Loss::CrossEntropy]
    }

    /// Build the loss into the tape and return the scalar loss handle.
    pub fn build(&self, tape: &mut Tape, output: Value, target: &Tensor) -> Value {
        match self {
            Loss::MeanSquaredError => tape.mse_loss(output, target.clone()),
            Loss::CrossEntropy => tape.softmax_cross_entropy(output, target.clone()),
        }
    }

    /// Eager scalar — handy for evaluating a model without the autograd tape.
    pub fn eval(&self, output: &Tensor, target: &Tensor) -> f32 {
        match self {
            Loss::MeanSquaredError => {
                assert_eq!(output.shape, target.shape, "MSE shape mismatch");
                let n = output.data.len() as f32;
                let mut s = 0.0_f32;
                for (a, b) in output.data.iter().zip(&target.data) {
                    let d = a - b;
                    s += d * d;
                }
                s / n
            }
            Loss::CrossEntropy => {
                assert_eq!(output.shape, target.shape, "CE shape mismatch");
                let sm = softmax_rows(output);
                let rows = sm.rows() as f32;
                let mut s = 0.0_f32;
                for (p, t) in sm.data.iter().zip(&target.data) {
                    s -= t * p.max(1e-12).ln();
                }
                s / rows
            }
        }
    }
}

File: dotnet/NeuralCabin.Engine/Loss.cs
namespace NeuralCabin.Engine;

/// <summary>Loss functions (port of <c>loss::Loss</c>).</summary>
public enum Loss
{
    MeanSquaredError,
    CrossEntropy,
}

public static class Losses
{
    public static string Name(this Loss l) => l switch
    {
        Loss.MeanSquaredError => "MeanSquaredError",
        Loss.CrossEntropy => "CrossEntropy",
        _ => throw new ArgumentOutOfRangeException(nameof(l)),
    };

    public static Loss Parse(string name) => name.ToLowerInvariant() switch
    {
        "mse" or "meansquarederror" => Loss.MeanSquaredError,
        "crossentropy" or "cross_entropy" => Loss.CrossEntropy,
        _ => throw new ArgumentException($"unknown loss '{name}'"),
    };

    /// <summary>
    /// Eager scalar loss over a batch.
    /// MSE: mean over all elements of (pred-target)².
    /// CE : mean over rows of −Σ_classes target·log(softmax(pred)).
    /// </summary>
    public static float Eval(this Loss l, Tensor output, Tensor target)
    {
        switch (l)
        {
            case Loss.MeanSquaredError:
            {
                var n = output.Length;
                var s = 0.0f;
                for (var i = 0; i < n; i++)
                {
                    var d = output.Data[i] - target.Data[i];
                    s += d * d;
                }
                return s / n;
            }
            case Loss.CrossEntropy:
            {
                var sm = Activations.SoftmaxRows(output);
                var rows = sm.Rows;
                var s = 0.0f;
                for (var i = 0; i < sm.Length; i++)
                    s -= target.Data[i] * (float)Math.Log(Math.Max(sm.Data[i], 1e-12f));
                return s / rows;
            }
            default:
                throw new ArgumentOutOfRangeException(nameof(l));
        }
    }

    /// <summary>
    /// Gradient of the loss w.r.t. the network's final output.
    /// For CrossEntropy the final output is the raw logits (the last activation,
    /// Softmax/Identity, is the identity during training — the loss folds in the
    /// softmax), giving (softmax(logits) − target)/rows.
    /// For MSE it is 2·(pred − target)/N over all elements.
    /// </summary>
    public static float[] Gradient(this Loss l, Tensor output, Tensor target)
    {
        var grad = new float[output.Length];
        switch (l)
        {
            case Loss.MeanSquaredError:
            {
                var n = output.Length;
                var scale = 2.0f / n;
                for (var i = 0; i < n; i++)
                    grad[i] = scale * (output.Data[i] - target.Data[i]);
                return grad;
            }
            case Loss.CrossEntropy:
            {
                var sm = Activations.SoftmaxRows(output);
                var rows = sm.Rows;
                var inv = 1.0f / rows;
                for (var i = 0; i < sm.Length; i++)
                    grad[i] = (sm.Data[i] - target.Data[i]) * inv;
                return grad;
            }
            default:
                throw new ArgumentOutOfRangeException(nameof(l));
        }
    }
}
